Add PSROIPooling Function

knorth55 opened the pull request:
Related to #538 
- [x] add `psroi_pooling_2d`
- [x] add test for `psroi_pooling_2d`
- [x] add doc for `psroi_pooling_2d`
- [x] add test to check `CPU` and `GPU` output is same.

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,331 @@
+    Args:
+        x (~chainer.Variable): Input variable. The shape is expected to be
+            4 dimentional: (n: batch, c: channel, h, height, w: width).
+        rois (~chainer.Variable): Input roi variable. The shape is expected to
```
I think it is not necessary to make the API of this function to follow `chainer.functions.roi_pooling_2d`.
Rather, it would be better to follow the preferred interface of ChainerCV.

Can you change the second input to `rois` and `roi_indices` as defined here? (`bbox` is `yx` order)
```
            rois (array): A bounding box array containing coordinates of
                proposal boxes.  This is a concatenation of bounding box
                arrays from multiple images in the batch.
                Its shape is :math:`(R', 4)`. Given :math:`R_i` proposed
                RoIs from the :math:`i` th image,
                :math:`R' = \\sum _{i=1} ^ N R_i`.
            roi_indices (array): An array containing indices of images to
                which bounding boxes correspond to. Its shape is :math:`(R',)`.
```
https://github.com/chainer/chainercv/blob/master/chainercv/links/model/faster_rcnn/faster_rcnn_vgg.py#L201

yuyu2172 replied:
This would require some change to kernels.

yuyu2172 commented on tests/functions_tests/test_psroi_pooling_2d.py:
```diff
@@ -0,0 +1,80 @@
+    def test_backward_gpu(self):
+        self.check_backward(cuda.to_gpu(self.x), cuda.to_gpu(self.rois),
+                            cuda.to_gpu(self.gy))
+
```
How about adding a test to check that CPU and GPU outputs are the same?


Related:
https://github.com/chainer/chainer/blob/master/tests/chainer_tests/functions_tests/array_tests/test_spatial_transformer_sampler.py#L119

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,331 @@
+    """Position Sensitive Region of Interest (ROI) pooling function.
+
+    This function computes position sensitive average of input spatial patch
+    with the given region of interest. Each ROI are splitted into
```
region of interest --> region of interests

Each ROI are --> Each ROI is

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,331 @@
+    This function computes position sensitive average of input spatial patch
+    with the given region of interest. Each ROI are splitted into
+    :math:`(group\_size, group\_size)` regions, and position sensitive values
+    in each region is inferenced.
```
infrenced --> computed

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,331 @@
+
+        bottom_data, bottom_rois = inputs
+        channels, height, width = bottom_data.shape[1:]
+        n_rois = bottom_rois.shape[0]
```
Could you follow ChainerCV's naming convention? (`n_{Singular}` instead of `n_{Plural}`)
`n_rois` --> `n_roi`

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,331 @@
+            roi_width = max(xmax - xmin, 0.1)
+            roi_height = max(ymax - ymin, 0.1)
+
+            strided = 1. * channels / self.outd
```
Could you remove `1.` and add `from __future__ import division`?

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,331 @@
+            roi_width = max(xmax - xmin, 0.1)
+            roi_height = max(ymax - ymin, 0.1)
+
+            strideh = 1. * roi_height / self.outh
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,331 @@
+        for i_roi in six.moves.range(n_rois):
+            idx, xmin, ymin, xmax, ymax = bottom_rois[i_roi]
+            idx = int(idx)
+            xmin = round(xmin * self.spatial_scale)
```
The variables are defined in `xy, wh` and `yx, hw` orders.
This is inconsistent. Could you reorder them to `yx, hw`?

knorth55 replied:
can it  be different from `roi_pooling_2d` in `chainer` master branch?

yuyu2172 replied:
Yes
https://github.com/chainer/chainercv/pull/545#discussion_r177318232

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,331 @@
+# Cuda Kernel Original work by Haozhi Qi (@Oh233)
```
Mention license

knorth55 commented:
@yuyu2172 I'm sorry that I rebase the commits again.
I follow your review and update.
Other thing is that I updated variables such as `outh` -> `out_h` and `sliceh` -> `slice_h`.
Also, I renamed `outd` -> `out_c`.

knorth55 commented on tests/functions_tests/test_psroi_pooling_2d.py:
```diff
@@ -0,0 +1,107 @@
+
+    @attr.gpu
+    @condition.retry(3)
+    def test_consistency_with_gpu(self):
```
I check that calculation output of GPU and CPU are same in this test.

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+            x_min = round(x_min * self.spatial_scale)
+            y_max = round(y_max * self.spatial_scale)
+            x_max = round(x_max * self.spatial_scale)
+            roi_width = max(x_max - x_min, 0.1)
```
Order of the two lines are opposite.

```
roi_height = max(y_max - y_min, 0.1)
roi_width = max(x_max - x_min, 0.1)
```

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+            'float32 top_data',
+            '''
+            // pos in output filter
+            int pw = i % pooled_width;
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+                round(bottom_rois[n * 4 + 3])) * spatial_scale;
+
+            // Force too small ROIs to be 1x1
+            float roi_width = max(roi_end_w - roi_start_w, 0.1);  // avoid 0
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+            bool is_empty = (hend <= hstart) || (wend <= wstart);
+
+            // Compute c at bottom
+            int gw = floor(
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+                static_cast<float>(pw) * group_size / pooled_width);
+            int gh = floor(
+                static_cast<float>(ph) * group_size / pooled_height);
+            gw = min(max(gw, 0), group_size - 1);
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+            x_min = round(x_min * self.spatial_scale)
+            y_max = round(y_max * self.spatial_scale)
+            x_max = round(x_max * self.spatial_scale)
+            roi_width = max(x_max - x_min, 0.1)
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+            ''',
+            'float32 top_diff',
+            '''
+            int pw = i % pooled_width;
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+                round(bottom_rois[n * 4 + 3])) * spatial_scale;
+
+            // Force too small ROIs to be 1x1
+            float roi_width = max(roi_end_w - roi_start_w, 0.1); // avoid 0
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+            float roi_height = max(roi_end_h - roi_start_h, 0.1);
+
+            // Compute w and h at bottom
+            float bin_size_w = roi_width / static_cast<float>(pooled_width);
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+            float bin_size_w = roi_width / static_cast<float>(pooled_width);
+            float bin_size_h = roi_height / static_cast<float>(pooled_height);
+
+            int wstart = floor(
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+                static_cast<float>(pw) * bin_size_w + roi_start_w);
+            int hstart = floor(
+                static_cast<float>(ph) * bin_size_h + roi_start_h);
+            int wend = ceil(
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+                static_cast<float>(ph + 1.0) * bin_size_h + roi_start_h);
+
+            // Add roi offsets and clip to input boundaries
+            wstart = min(max(wstart, 0), width);
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+            // Add roi offsets and clip to input boundaries
+            wstart = min(max(wstart, 0), width);
+            hstart = min(max(hstart, 0), height);
+            wend = min(max(wend, 0), width);
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+            bool is_empty = (hend <= hstart) || (wend <= wstart);
+
+            // Compute c at bottom
+            int gw = floor(
```
ditto

yuyu2172 commented on chainercv/functions/psroi_pooling_2d.py:
```diff
@@ -0,0 +1,368 @@
+                static_cast<float>(pw) * group_size / pooled_width);
+            int gh = floor(
+                static_cast<float>(ph) * group_size / pooled_height);
+            gw = min(max(gw, 0), group_size - 1);
```
ditto
